### DISCUSSION TITLE
How to communicate with developers better?

### DISCUSSION THREAD

**Discussion by LaserEyess** [transmission/transmission#2414] [General]
First, I want to say I'm really glad to see a lot of activity from core developers in the past few months, especially after concerns that the project was dying. However, I still notice a strange disconnect between what a) users want and contributors are doing, and b) what developers seem to care about. When I asked in `#transmission` on irc.libera.chat, I was told discussions were the place to get in touch with developers. But I'd like to talk about things like:

 - What features are the devs interested in long term, i.e. what would they be willing to both review and merge? Do users have a voice in this?
 - I see milestones as well as labels like "pr welcome"/"help wanted" but how are these chosen?
 - Is there a long term maintenance plan or is it just a "we'll work on this when we have time and then if we don't have time we won't" type of thing?
 - Are there any plans to add more people with merge access to the repo?

TL;DR I guess this project seems somewhat opaque at times, I have no idea what developers discuss behind the scenes, or if there is any discussion at all rather than individuals acting individually. Maybe that's by design but I think it would be nice for more openness.

**Comment by ckerr** [comment]
> What features are the devs interested in long term

I've been writing / landing most of the changes lately. My personal wishlist is:

- Support the updated BitTorrent spec, e.g. the new .torrent file format, merkle tree requests, etc.
- A more efficient RPC REST API. This would take some consideration, as there are a _lot_ of RPC clients out there so we would want to avoid ripple effects that break those clients.
- Patches that improve transfer speeds
- Patches that improve communication with peers, indirectly improving transfer speeds
- Patches that reduce memory overhead
- Patches that reduce CPU overhead
- Patches that simplify the code, e.g. reduce code size or complexity.
- Patches that improve testability, e.g. decouple pieces of code that can't currently be tested in isolation due to interdependencies. This is the biggest reason that testing isn't better in libtransmission. (Related to previous line)
- **TESTS!**  (Related to previous line) especially in libtransmission or in the web client
- Fix deprecations / warnings in the macOS client
- Patches that increase C++ modernization, especially if that advances other goals -- e.g. replace raw pointers with std::shared_ptr or std::unique_ptr (reducing code complexity), or which replace tr_ptrArray with std::vector (which would let us remove tr_ptrArray, reducing code size).

To make the list a little longer -- a couple of things I probably wouldn't write but wish someone else would:

- Refactor the GTK client to be more like the Qt client, e.g. add the ability to act as a remote control to transmission-daemon
- Update the GTK client to fix all deprecated glib/GTK API use
- Update the GTK client to better follow the Gnome HIG (but without losing some of the existing "power user" features -- the mockups I've seen in the past all seem to assume users never have more than a few torrents at once)
- Rewrite both the GTK and macOS clients so to use RPC so that we can kill of the public C API :laughing: 
 
> i.e. what would they be willing to both review and merge?

Some guidelines:

- Patches that advance any of the goals in the previous checklist would be _very_ welcome
- One of Transmission's biggest strengths is native UIs. This is also a weakness in terms of development, because any UI-visible feature has to be added to multiple clients. So:
  - If a feature isn't OS-specific, prefer to add code to `libtransmission/` and expose it via RPC and via the public C API so that it can be used by all the clients
  - For rarely-used / power user features, consider not having a UI at all but rather exposing it in the config file, so that all the UIs don't have to be updated
> Do users have a voice in this?

Of course! PRs are always welcomed. Is there a reason why you think they wouldn't be -- are there any PRs that you think I've incorrectly turned down?

> Is there a long term maintenance plan or is it just a "we'll work on this when we have time and then if we don't have time we won't" type of thing?

Nobody's earning a paycheck to work on Transmission, so there's no commitment to timelines or release dates, no. 

> Are there any plans to add more people with merge access to the repo?

I'm open to the idea -- the project has a very high bus factor right now. Progress on the project more-or-less mirrors my own personal interest in the project, which has varied over the years.

Note, adding write permissions is a significantly greater amount of trust than merging someone's PR. I'd welcome PRs from anyone, but there would need to be a significant track record of merged PRs before write permissions were granted. At the moment the only person who even halfway matches that criteria is @nevack, though (a) we haven't discussed it / I don't know if he'd be interested, and (b) I'd ideally like to see more commits and a longer tracker record.

That's not saying anything against nevack or any other contributor. PRs are absolutely welcomed.

> TL;DR I guess this project seems somewhat opaque at times, I have no idea what developers discuss behind the scenes, or if there is any discussion at all rather than individuals acting individually. Maybe that's by design but I think it would be nice for more openness.

What kind of openness would you suggest?

I don't want to commit to anything that adds a time commitment in the long term, e.g. idling in irc. But I'd be happy to things that don't require a time commitment, e.g. updates to the README to give more information (like the stuff written above) to potential contributors.

**Reply by nevack** [reply]
Hi, as I was mentioned here, wanna leave some thoughts of mine.

I definitely want to help with the project, but I have a full-time job and can't promise to make a lot of improvements.

If you need more eyes on PRs for macOS client, you can always add me as a reviewer.
It's important for the community to receive feedback and iterate fast, to make the ideal solution and deliver it to others.

What about the write access to the repo - I don't want it and I don't mind it.
We can discuss it later when you think it's the right time.

It doesn't bother me to make PRs without being able to merge them by myself, I think it's important to be peer-reviewed and have external approves. Mistakes could affect a lot of users, macOS nightly builds are used extensively AFAIK.

What bothers me is undefined review periods and being unable to communicate with code-owners.

Slack/Discord server will definitely help with the latter.

**Comment by LaserEyess** [comment]
Thank you for the well thought-out reply, I really appreciate that. I think that's a great list of things to focus on, especially transfer speed and peering communication. I see issues like [this](https://github.com/transmission/transmission/issues/534) or [this](https://github.com/transmission/transmission/issues/1314) which look really tough to fix for newer developers who don't know the architecture of the code well. Hence, why support from core developers is needed. As you say the bus factor is high and yet no one is getting paid to work on this so it is at the whims of your motivations. I hope that doesn't sound too ungrateful but from my perspective it is a precarious position to be in for software I rely upon. 

I do hope to contribute some to transmission proper, I'm cleaning up a transmission client [transmission-remote-gtk](https://github.com/transmission-remote-gtk/transmission-remote-gtk) and I'd like to contribute not only as a user of the daemon, but as a developer of an RPC client. 

> Of course! PRs are always welcomed. Is there a reason why you think they wouldn't be -- are there any PRs that you think I've incorrectly turned down?

I can point to a specific example, actually: [this PR](https://github.com/transmission/transmission/pull/1757). Specifically this one because I feel it is a feature you might not consider useful at all, or even dangerous, and don't know that it's a very common use case. Creating torrents locally to add to transmission is something I commonly do, and I have to sit there and wait for the torrent I created with `transmission-create` to be verified in `transmission-daemon`, despite the fact that the daemon should be able to trust the files exist, because how else would I have made the torrent? 

But maybe you don't know this is a common use case for users, or maybe you do and you still do not think it's a good feature. Regardless, I don't see any place for people to argue their usecase except in the PR itself, which was abandoned. Right now I have assumed silence means your mind has been made up, but I don't know. Your mind has been [changed](https://github.com/transmission/transmission/issues/597#issuecomment-945145569) on at least one feature after examining the bittorrent ecosystem and concluding that this feature makes more sense in the current ecosystem.

That was a bit of an aside but just an anecdote about feedback from developers I sometimes see missing.

>What kind of openness would you suggest?
>
>I don't want to commit to anything that adds a time commitment in the long term, e.g. idling in irc. But I'd be happy to things that don't require a time commitment, e.g. updates to the README to give more information (like the stuff written above) to potential contributors.

I would have suggested a `#transmission-devel` IRC channel, though i obviously know `#transmission` exists. Or a discord, or slack, or whatever else exists for instant communication. I would not consider that a time commitment as idling, but if that's not your preferred method of communication it can't be helped. Discussions are one option, but maybe I should give some concrete examples of things I'd like to ask developers.

 - What are all the places I need to change in order to add a new RPC option, such as [this one](https://github.com/transmission/transmission/issues/1656) or adding labels to the `torrent-add` method.
 - Why is there excessive locking on accessing the session/torrent strucks on RPC/disk I/O? Can these be refactored to remove things [like this?](https://github.com/transmission/transmission/issues/1540) (I also suspect a lot of performance issues are caused by this)
 - What refactoring would be needed to support HTTP/2 for RPC, or even a unix socket? E.g., what else in [this PR](https://github.com/transmission/transmission/pull/1673)
 - What is the logic behind the internal architecture of transmission? Why are things threaded like they are, etc.

Those are pretty big, broad questions, I know, but for someone sitting down and trying to look into what to do, being able to actually *talk* to someone about it would be *extremely* helpful. Ultimately that's the type of back and forth I would be interested in. 

Though this part, specifically: 
>But I'd be happy to things that don't require a time commitment, e.g. updates to the README to give more information (like the stuff written above) to potential contributors.

I think this would be a great addition.

**Comment by ckerr** [comment]
> Creating torrents locally to add to transmission is something I commonly do, and I have to sit there and wait for the torrent I created with transmission-create to be verified in transmission-daemon, despite the fact that the daemon should be able to trust the files exist, because how else would I have made the torrent?

I _really_ dislike having a "skip verify" feature due to the obvious potentials for abuse, but maybe there is a way we can both get what we want, at least in `main`. The resume file now holds progress [this way](https://github.com/transmission/transmission/blob/f415de0626cd88ad9b4cbedba10c53dfad0a24c8/libtransmission/resume.cc#L489):

>  * - 'pieces' a bitfield for whether each piece has been checked.
>  * - 'mtimes', an array of per-file timestamps

Transmission, when loading up a torrent, checks to see if the files' mtimes on disk match the ones in the `mtimes` array. If the file has changed, the corresponding piece flags in the bitfield are zeroed out, indicating that they need to be checked before use.

What if Transmission's torrent-add RPC method let you pass in this info? Effectively you'd be saying "here are the mtimes that the files had when I scanned them all to create the .torrent file." That way Transmission would be able to compare times, and if they were the same (which they would be, since you _just created the .torrent_), would skip verifying that piece without adding in a brute force "[x] skip verify" checkbox.

There would be some other changes needed in the torrent bootstrap code, e.g. avoiding the call to tr_torrentVerify() iff all the mtimes are  up-to-date, but the basic idea would work.

If this would interest you, please open a new issue (or PR! :smile_cat:) and we can continue discussion there.

---  

> Discussions are one option, but maybe I should give some concrete examples of things I'd like to ask developers.

Discord or Slack might work if people are interested? TBH I have a slight preference for discussions, since that leaves a web-searchable paper trail that might be of use for other would-be contributors. IRC seems worst of all to me for that same reason.

---

> What are all the places I need to change in order to add a new RPC option, such as this one or adding labels to the torrent-add method.

First step / easiest effort: file a new issue or PR just changing `extras/rpc-spec.txt` so that we can discuss the API itself

Implementation: GTK and macOS clients still use the C API, so new features in libtransmission/ need to be exposed via both `libtransmission/transmission.h` and via RPC. So

- extras/rpc-spec.txt
- libtransmission/rpcimpl.cc
- libtransmission/transmission.h
- libtransmission/ files relevant to implementing the change 
- if the change involves new UI, please add that UI in as many of the clients as possible. Full coverage is not required, but please at least cover two UIs.

---

> Why is there excessive locking on accessing the session/torrent strucks on RPC/disk I/O? Can these be refactored to remove things like this? (I also suspect a lot of performance issues are caused by this)

Probably because the libtransmission thread blocks on IO, which is a major wart. There are some advantages, e.g. there are no edge cases / race conditions when removing a torrent or when a torrent becomes complete, since computation and disk IO are synchronous. The disadvantage is obvious. The least-complex refactor would probably be to delegate disk IO out to workers in a threadpool that then report back to the libtransmission thread when the work is done.

See also: https://github.com/transmission/transmission/issues/108, https://github.com/transmission/transmission/issues/1088,  https://github.com/transmission/transmission/issues/2011

---

> What refactoring would be needed to support HTTP/2 for RPC, or even a unix socket? E.g., what else in this PR

Right now Transmission uses libevent's built-in HTTP server for RPC. In order to support more options we would probably want to consider switching to a different solution. The recent C++ switch probably makes this easier since it widens our options. My preference would be a server that is actively maintained, narrow in scope, preferably few LOC / few dependencies. I'd love to see Transmission server RPC with https instead of http. If anyone wants to research the different options down there and try to find the projects that fit best, that would be wonderful

---  

> What is the logic behind the internal architecture of transmission? Why are things threaded like they are, etc.

This is too large of a question to give a useful answer. Probably better to file other discussion questions with more narrowly-focused questions. :smile: 

This discussion is branching out into at least five separate discussions. Just for thread management, we should probably abandon / lock this discussion and follow up individual topics in other discussions or issues or PRs.

Also I've opened up https://github.com/transmission/transmission/issues/2428 as a TODO item for updating the README with some of the ideas discussed above

**Comment by LaserEyess** [comment]
> This is too large of a question to give a useful answer. Probably better to file other discussion questions with more narrowly-focused questions.

I think a lot of the questions I asked were, yes. The trouble with things like github discussions is that having long conversations like this tend to fragment discussions like there where now we have 3-4 different trains of thought we're keeping track of in this single discussion. I guess that warrants a discussion thread for each of them, but that sort of fragmentation is what makes conversations hard to follow. I do see your point about chatting logs being lost to time, though there are ways to prevent that from happening if you're truly determined.

I think I will make separate discussion threads or issues where relevant. Thank you for your time and your responses, I'm really glad to see transmission development picking back up. This is one of the only clients that treats RPC and remote management as first class.

**Comment by ckerr** [comment]
> As you say the bus factor is high and yet no one is getting paid to work on this so it is at the whims of your motivations. I hope that doesn't sound too ungrateful but from my perspective it is a precarious position to be in for software I rely upon.

The obvious solution to this would be for you to start submitting patches :green_heart: as it would give you more of a stake and reduce the bus factor at the same time.

**Reply by GaryElshaw** [reply]
Not being unusually demanding and mean to the devs would probably be a good move too, yeah?

**Reply by LaserEyess** [reply]
> The obvious solution to this would be for you to start submitting patches as it would give you more of a stake and reduce the bus factor at the same time.

I think maybe I didn't make my point clear enough earlier about this part. But nevack put it better above. It's not about "giving people write access" just to give them a special status. The issue is what happens when the developers no longer have any interest and there's no one who can merge PRs *at all*? This was the case with transmission for a few years from like 2018-2021. Many people *did* submit PRs and they *did* try to actively contribute, but it's hard to do so when the other side gives nothing but silence. A lot of these PRs are abandoned and their authors uninterested anymore potentially because they thought the devs either did not care, or that the project was abandoned.

I will certainly try to contribute, especially while there are people around to review and merge PRs I make, but the overarching point I wanted to make was that the devs communicating better would be a better way to attract and maintain consistent contributors. Again, even today you are making huge strides in closing that gap, so please don't think this is an active criticism of what you're doing.

**Comment by ckerr** [comment]
https://github.com/transmission/transmission/pull/2452 has a new `CONTRIBUTING.md` that folds in some of the guidelines from the previous `libtransmission/README.md` and some of the comments from upthread here.

I'd love it if the people here would do a review on it. Attracting new contributors would be a healthy thing for the project!

**Comment by rpatterson** [comment]
I made an attempt to [divert off topic discussion out of an issue](https://github.com/transmission/transmission/issues/7078#issuecomment-2463132504) and I linked here for an example. I hope that's ok.

I'm also going take this opportunity to speak my peace here without adding to that off-topic noise. I'm so grateful to the maintainers. Threatening or shaming volunteers in open source communities is counterproductive at best and abusive at worst. I do think that v4.0.6 bug and the delays in releasing a fix reflect poorly on this project to those outside this community but the solution to do that is to step up and do something and I'm not going to so I've got nothing to say. Except that. ;-)

**Comment by Coeur** [comment]
>Note, adding write permissions is a significantly greater amount of trust than merging someone's PR. I'd welcome PRs from anyone, but there would need to be a significant track record of merged PRs before write permissions were granted. At the moment the only person who even halfway matches that criteria is @nevack, though (a) we haven't discussed it / I don't know if he'd be interested, and (b) I'd ideally like to see more commits and a longer tracker record.

Hi @ckerr. I missed that discussion before. Could you go about pro-actively recontact nevack or tearfur for their interest to get write access, in order to get the extra help that the project needs regarding maintaining that huge backlog of 100 opened pull requests?

**Comment by LaserEyess** [comment]
It looks like this is getting attention again, and after ~2.5+ years it seems that we are again in a situation where communication with the devs is difficult and sporadic. I think these questions are still relevant:

- Is there a long term maintenance plan or is it just a "we'll work on this when we have time and then if we don't have time we won't" type of thing?
- Are there any plans to add more people with merge access to the repo?  (This is about reducing the bus factor, not about a status symbol)

And I want to bring this quote up again:

> I think maybe I didn't make my point clear enough earlier about this part. But nevack put it better above. It's not about "giving people write access" just to give them a special status. The issue is what happens when the developers no longer have any interest and there's no one who can merge PRs at all? This was the case with transmission for a few years from like 2018-2021. **Many people did submit PRs and they did try to actively contribute, but it's hard to do so when the other side gives nothing but silence. A lot of these PRs are abandoned and their authors uninterested anymore potentially because they thought the devs either did not care, or that the project was abandoned.**

This is what I see happening again. And I'm hesitant myself to contribute more to transmission because I do not know if I'm wasting my time. Even if I tackled a "pr welcome" issue, is it going to be merged? Do I have to endlessly wait for it because it's tagged with 4.1 or 4.x and it can't be merged until 4.0.7 is out? Will it even get reviewed? I think this is weighing on other contributors as well.
